eppo_core/src/events/event_delivery.rs

```diff
@@ -0,0 +1,62 @@
+use crate::events::event::Event;
+use crate::Error;
+use log::{debug, info};
+use reqwest::{StatusCode};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+#[derive(Clone)]
+pub struct EventDelivery {
+    sdk_key: String,
+    ingestion_url: String,
+    client: reqwest::Client,
+}
+
+#[derive(serde::Deserialize)]
+pub struct EventDeliveryResponse {
+    pub failed_events: Vec<Uuid>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct IngestionRequestBody {
+    eppo_events: Vec<Event>,
+}
+
+/// Responsible for delivering event batches to the Eppo ingestion service.
+impl EventDelivery {
+    pub fn new(sdk_key: String, ingestion_url: String) -> Self {
+        let client = reqwest::Client::new();
+        EventDelivery {
+            sdk_key,
+            ingestion_url,
+            client,
+        }
+    }
+
+    // Delivers the provided event batch and returns a Vec with the events that failed to be delivered.
+    pub async fn deliver(&self, events: Vec<Event>) -> Result<EventDeliveryResponse, Error> {
+        let ingestion_url = self.ingestion_url.clone();
+        let sdk_key = self.sdk_key.clone();
+        debug!("Delivering {} events to {}", events.len(), ingestion_url);
+        let body = IngestionRequestBody { eppo_events: events };
+        let response = self.client.post(ingestion_url)
+            .header("X-Eppo-Token", sdk_key)
+            .json(&body)
+            .send()
+            .await?;
+        let response = response.error_for_status().map_err(|err| {
+            return if err.status() == Some(StatusCode::UNAUTHORIZED) {
+                // TODO: Mark all events as failed, indicate that this error is not-retriable
+                log::warn!(target: "eppo", "client is not authorized. Check your API key");
+                Error::Unauthorized
```

Comment:
good catch about unauthorized!

minor: For configuration polling, that usually means that SDK key is not valid, so we just stop the poller. Does it makes sense to do the same here? If we failed to send one batch because we're unauthorized, sending the next batch would likely fail in the same way

Reply:
yeah exactly the same!